test/irmin/fuzz.ml

```diff
@@ -0,0 +1,147 @@
+              in
+              Store.Tree.list tree [] >>= fun dir ->
+              let dir_len = List.length dir in
+              (* We use [randoms] to shallow differently the siblings within [dir]
+                 and we pass a tail of [randoms] in recursive calls, to have
+                 variance in subtrees. *)
+              let dir' = combine_drop dir (enlarge dir_len randoms) in
+              assert (List.length dir' = dir_len);
+              Lwt_list.fold_left_s
+                (fun (shallowed, acc) ((k, subtree), i) ->
+                  (* random oracle to decide whether to make complety shallow
+                     (b holds) or partially shallow (b doesn't hold: recurse) *)
+                  let b = i mod 2 = 0 in
+                  (if b then make_tree_shallow repo subtree |> Lwt.return_some
+                  else
+                    let sub_randoms =
+                      match randoms with
+                      | [] -> assert false
+                      | [ x ] -> [ x ]
+                      | _ :: rest -> rest
+                    in
+                    make_partially_shallow repo subtree sub_randoms)
+                  >>= fun subtree_opt ->
+                  let shallowed', subtree' =
+                    match subtree_opt with
+                    | None -> (shallowed, subtree) (* no change *)
+                    | Some subtree' ->
+                        (* subtree' is partially shallow *)
+                        (true, subtree')
+                  in
+                  Store.Tree.add_tree acc [ k ] subtree' >>= fun acc ->
+                  Lwt.return (shallowed', acc))
+                (false, Store.Tree.empty) dir'
+              >>= fun (shallowed, tree') ->
+              if shallowed then Lwt.return_some tree' else Lwt.return_none)
+
+  (** Test that subtituting subtrees by their [Store.Tree.shallow] version
+      doesn't change the tree's top-level hash. *)
+  let test_hash_stability repo tree is =
+    make_partially_shallow repo tree is >>= fun tree_opt ->
+    match tree_opt with
+    | None -> Crowbar.bad_test ()
+    | Some tree' ->
+        let hash_to_string = Irmin.Type.to_string Store.Hash.t in
+        let hash_eq = Irmin.Type.(unstage (equal Store.Hash.t)) in
+        let hash = Store.Tree.hash tree in
+        let hash' = Store.Tree.hash tree' in
+        let err_opt =
+          if hash_eq hash hash' then None
+          else
+            Some
+              (Printf.sprintf "Hash mismatch: %s <> %s\n" (hash_to_string hash)
+                 (hash_to_string hash'))
+        in
+        (match err_opt with None -> () | Some err -> Crowbar.fail err);
+        Lwt.return_unit
+
+  let test_hash_stability r t is = test_hash_stability r t is |> Lwt_main.run
+end
+
+let () =
+  Crowbar.add_test
+    ~name:"Store.Tree.hash t = Store.Tree.hash (make_partially_shallow t)"
+    [ Generators.irmin_tree_gen; Crowbar.(list1 int) ]
```

Comment:
Instead of a relying on a list of integers for randomness, you could generate a single integer here, and use it as a seed to spawn a `Random.State`. It would then allow you to simplify a lot `make_partially_shallow`.

Reply:
I haven't looked into the details, but I'm concerned that this would end up offloading the work of determining the _length_ of the list (which is a proxy for the tree size) from Crowbar. That's potentially an issue because e.g. it prevents Crowbar from generating so-called "small examples".

Reply:
`Generators.irmin_tree_gen` is the one in charge of generating the structure of the tree. `Crowbar.(list1 int)` is here to decide which branch of the tree should be shallowed.

Reply:
@Ngoguey42> yes you're correct, `list1 int` is used as randomizing data. And yes, it would be better to spawn a `Random.State` instead. I'm OOO until March 1st and won't do any change until then, feel free to do improvements :+1:

Reply:
I implemented your suggestion in 104fa38

Reply:
Thank you!